storage/replica_proposal.go

```diff
@@ -431,37 +490,51 @@ func (r *Replica) handleTrigger(
 		}); err != nil {
 			log.Infof(ctx, "unable to gossip first range: %s", err)
 		}
+		pd.gossipFirstRange = false
 	}
 
-	if trigger.addToReplicaGCQueue {
+	if pd.addToReplicaGCQueue {
 		if _, err := r.store.replicaGCQueue.Add(r, replicaGCPriorityRemoved); err != nil {
 			// Log the error; the range should still be GC'd eventually.
 			log.Errorf(ctx, "unable to add to replica GC queue: %s", err)
 		}
+		pd.addToReplicaGCQueue = false
 	}
 
-	if trigger.maybeAddToSplitQueue {
+	if pd.maybeAddToSplitQueue {
 		r.store.splitQueue.MaybeAdd(r, r.store.Clock().Now())
+		pd.maybeAddToSplitQueue = false
 	}
 
-	if trigger.maybeGossipSystemConfig {
+	if pd.maybeGossipSystemConfig {
 		r.maybeGossipSystemConfig()
+		pd.maybeGossipSystemConfig = false
 	}
 
-	// On the replica on which this command originated, resolve skipped intents
-	// asynchronously - even on failure.
-	//
-	// TODO(tschottdorf): EndTransaction will use this pathway to return
-	// intents which should immediately be resolved. However, there's
-	// a slight chance that an error between the origin of that intents
-	// slice and here still results in that intent slice arriving here
-	// without the EndTransaction having committed. We should clearly
-	// separate the part of the trigger which also applies on errors.
 	if originReplica.StoreID == r.store.StoreID() {
-		r.store.intentResolver.processIntentsAsync(r, trigger.intents)
+		// On the replica on which this command originated, resolve skipped
+		// intents asynchronously - even on failure.
+		//
+		// TODO(tschottdorf): EndTransaction will use this pathway to return
+		// intents which should immediately be resolved. However, there's
+		// a slight chance that an error between the origin of that intents
+		// slice and here still results in that intent slice arriving here
+		// without the EndTransaction having committed. We should clearly
+		// separate the part of the ProposalData which also applies on errors.
+		r.store.intentResolver.processIntentsAsync(r, pd.intents)
+		if pd.leaseMetricsResult != nil {
+			r.store.metrics.leaseRequestComplete(*pd.leaseMetricsResult)
+		}
+	}
+	pd.leaseMetricsResult = nil
+	pd.intents = nil
+
+	if pd.ComputeChecksum != nil {
+		r.computeChecksumPostApply(ctx, *pd.ComputeChecksum)
+		pd.ComputeChecksum = nil
 	}
 
-	if trigger.computeChecksum != nil {
-		r.computeChecksumTrigger(ctx, *trigger.computeChecksum)
+	if !reflect.DeepEqual(pd, ProposalData{}) {
```

Comment:
I'll clean this up. I'll also make sure that we don't always call `r.assertState` before this merges (which is done elsewhere).